daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Classify.scala

```diff
@@ -85,7 +83,9 @@ private[speedy] object Classify { // classify the machine state w.r.t what step
       classifyKont(kont, counts)
     } else {
       counts.ctrlExpr += 1
-      classifyExpr(machine.ctrl, counts)
+      if (machine.ctrl != null) {
```

Comment:
Don't we preserve the invariant that exactly one of `machine.returnValue` and `machine.ctrl` is `null`? If so, an `assert(machine.ctrl != null)` is probably better.

Reply:
We do. But if the invariant is broken, the code will crash in speedy `run()` on this line: `expr.execute(this)`. It is confusing if we crash in a different place when `enableInstrumentation` is true.

Reply:
I understand. Thanks.